arrow-core-data/src/test/kotlin/arrow/core/NullableTest.kt

```diff
@@ -0,0 +1,187 @@
+@file:Suppress("NAME_SHADOWING")
+package arrow.core
+
+import io.kotlintest.matchers.collections.shouldContainAll
+import io.kotlintest.matchers.types.shouldBeNull
+import io.kotlintest.matchers.types.shouldNotBeNull
+import io.kotlintest.shouldBe
+import io.kotlintest.specs.StringSpec
+
+/*
+----------------------------------------------------------------------------------------------------
+ */
+
+class NullableTest : StringSpec({
+  "map1 short circuits if any arg is null" {
+    map1(null) { _ -> Unit }.shouldBeNull()
+  }
+
+  "map1 performs action when arg is not null" {
+    map1(1) { a -> a + 1 }.let {
+      it.shouldNotBeNull()
+      it shouldBe 2
+    }
+  }
+
+  "map2 only performs action when all arguments are not null" {
+    generateAllPathsForNForks("a", null, 2)
+      .forEach { (a: String?, b: String?) ->
+        if (listOf(a, b).all { it != null }) {
+          map2(a, b, { a, b -> a + b }).let {
+            it.shouldNotBeNull()
+            it shouldBe a!! + b!!
+          }
+        } else {
+          map2(a, b, { _, _ -> Unit }).shouldBeNull()
+        }
+      }
```

Comment:
Here and below: in order to use property based testing like in other tests, we should use `forAll` with a `Gen<A>`. The good news is that you already have a very good data generator, so you only need to convert it to a gen, as follows:

```
private fun <A> combGen(choice1: A, choice2: A, n: Int) = object : Gen<List<A>> {
  override fun constants(): Iterable<List<A>> = emptyList()

  override fun random(): Sequence<List<A>> = generateAllPathsForNForks(choice1, choice2, n).asSequence()
}
```

with this, you can then rewrite the tests as follows:

```suggestion
    forAll(combGen("a", null, 2)) { (a: String?, b: String?) ->
      if (listOf(a, b).all { it != null }) {
        map2(a, b, { a, b -> a + b }).let {
          it == a!! + b!!
        }
      } else {
        map2(a, b, { _, _ -> Unit }) == null
      }
    }
```

essentially just replacing `generateAllPathsForNForks(???).forEach` with `forAll(combGen(???))` 👌